### PR TITLE
Fix write_trailer crash by checking if header is written

### PR DESCRIPTION
**Pull request by golgetahir**
This PR fixes #https://github.com/ant-media/Ant-Media-Server/issues/3201

### CODE REVIEW

File: src/main/java/io/antmedia/muxer/RtmpMuxer.java [hunk 1]
```diff
@@ -199,13 +200,12 @@ public  boolean prepareIO()
...
-	 * writeHeader and writeTrailer methods are synchronized. 
-	 * Because we have encountered some cases that while it's in writeHeader, writeTrailer is called. 
+	 * If the broadcast is stopped while the muxer is writing the header
+	 * it cannot complete writing the header
 	 * Then writeTrailer causes crash because of memory problem.
-	 * 
-	 * synchronized methods are not called at the same time from different threads
+	 * We need to control if header is written before trying to write Trailer.
 	 */
-	private synchronized boolean writeHeader() {
```
Comment (mekya) [inline src/main/java/io/antmedia/muxer/RtmpMuxer.java:208]: Why do we remove `synchronized`  in writeHeader and writeTrailer ?

They may enter the both methods at the same time.

File: src/main/java/io/antmedia/muxer/RtmpMuxer.java [hunk 2]
```diff
@@ -199,13 +200,12 @@ public  boolean prepareIO()
...
+	private boolean writeHeader() {
```
Comment (mekya) [inline src/main/java/io/antmedia/muxer/RtmpMuxer.java:208]: I think we also need to include this case. `writeHeader` may be called after `writeTrailer`.
For this case, if `writeTrailer`is callled, we do nothing in the `writeHeader`

**Review by mekya — CHANGES_REQUESTED** [review]

File: src/main/java/io/antmedia/muxer/RtmpMuxer.java (hunk 1 above)
Comment (golgetahir) [reply src/main/java/io/antmedia/muxer/RtmpMuxer.java:208]: It is done

File: src/main/java/io/antmedia/muxer/RtmpMuxer.java (hunk 2 above)
Comment (golgetahir) [reply src/main/java/io/antmedia/muxer/RtmpMuxer.java]: I have added a trailerWritten volatile boolean for synchronization

**Comment by mekya** [comment]
Is there a test code for this issue?

**Review by mekya — APPROVED** [review]

**Pull request merged**